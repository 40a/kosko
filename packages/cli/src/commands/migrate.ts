import { migrateString } from "@kosko/migrate";
import fs from "fs";
import getStdin from "get-stdin";
import { join, resolve } from "path";
import { promisify } from "util";
import { Command, RootArguments } from "../cli/command";
import Debug from "../cli/debug";
import { print } from "../cli/print";

const debug = Debug.extend("migrate");
const stat = promisify(fs.stat);
const readFile = promisify(fs.readFile);
const readDir = promisify(fs.readdir);

function readFileString(path: string) {
  debug("Reading file", path);
  return readFile(path, "utf8");
}

async function readFilesInDir(dir: string) {
  debug("Reading directory", dir);

  const files = await readDir(dir);
  const contents = await Promise.all(
    files.map(file => readFileString(join(dir, file)))
  );

  return concatFiles(contents);
}

function concatFiles(arr: ReadonlyArray<string>): string {
  if (!arr.length) return "";
  let output = "";

  for (const s of arr) {
    if (!s.startsWith("---")) output += "---\n";
    output += s + "\n";
  }

  return output;
}

function readFiles(
  cwd: string,
  files: ReadonlyArray<string>
): Promise<ReadonlyArray<string>> {
  return Promise.all(
    files.map(async file => {
      if (file === "-") {
        debug("Reading from stdin");
        return getStdin();
      }

      const path = resolve(cwd, file);
      const stats = await stat(path);

      return stats.isDirectory() ? readFilesInDir(path) : readFileString(path);
    })
  );
}

export interface MigrateArguments extends RootArguments {
  filename: string[];
}

export const migrateCmd: Command<MigrateArguments> = {
  command: "migrate",
  describe: "Migrate YAML into components",
  builder(argv) {
    /* istanbul ignore next */
    return argv.option("filename", {
      type: "string",
      describe: "Files, directory or stdin to migrate",
      required: true,
      alias: "f",
      array: true
    });
  },
  async handler(args) {
    const file = concatFiles(await readFiles(args.cwd, args.filename));
    const content = migrateString(file);

    await print(content);
  }
};
