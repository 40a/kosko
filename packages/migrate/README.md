# @kosko/migrate

[Docs](https://github.com/tommy351/kosko)

## API

Migrate Kubernetes manifests into a kosko component.

### `migrate`

```js
migrate(data: ReadonlyArray<Manifest>): string
```

### `migrateString`

Migrate Kubernetes YAML into a kosko component.

```js
migrateString(input: string): string
```
