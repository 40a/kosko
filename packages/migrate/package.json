{
  "name": "@kosko/migrate",
  "version": "0.0.1",
  "description": "Migrate Kubernetes YAML into kosko components.",
  "homepage": "https://github.com/tommy351/kosko",
  "repository": {
    "type": "git",
    "url": "https://github.com/tommy351/kosko.git"
  },
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist",
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">= 6.9.0"
  },
  "dependencies": {
    "camelcase": "^5.0.0",
    "js-yaml": "^3.12.1",
    "kubernetes-models": "^0.4.0",
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@types/camelcase": "^4.1.0",
    "@types/js-yaml": "^3.12.0"
  }
}
