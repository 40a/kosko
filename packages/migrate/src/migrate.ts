import yaml from "js-yaml";
import camelcase from "camelcase";
import { getImportPath } from "kubernetes-models/resolve";

export interface Manifest {
  apiVersion: string;
  kind: string;
  [key: string]: any;
}

interface Component {
  name: string;
  text: string;
  imports: ReadonlyArray<Import>;
}

interface Import {
  names: ReadonlyArray<string>;
  path: string;
}

function jsonStringify(data: any): string {
  return JSON.stringify(data, null, "  ");
}

function generateComponent(manifest: Manifest): Component {
  const { apiVersion, kind, ...data } = manifest;
  const name = camelcase(kind);
  const importPath = getImportPath(apiVersion, kind);

  if (!importPath) {
    return {
      name,
      text: jsonStringify(manifest),
      imports: []
    };
  }

  return {
    name,
    text: `new ${kind}(${jsonStringify(data)})`,
    imports: [
      {
        path: "kubernetes-models/" + importPath,
        names: [kind]
      }
    ]
  };
}

function validateManifest(data: any): data is Manifest {
  return typeof data.apiVersion === "string" && typeof data.kind === "string";
}

function generateForList(
  items: ReadonlyArray<unknown>
): ReadonlyArray<Component> {
  return items
    .map(data => {
      if (!validateManifest(data)) {
        throw new Error("apiVersion and kind is required");
      }
      if (data.apiVersion === "v1" && data.kind === "List") {
        return generateForList(data.items);
      }

      return generateComponent(data);
    })
    .reduce((acc: Component[], x) => acc.concat(x), []);
}

function uniqComponentName(
  components: ReadonlyArray<Component>
): ReadonlyArray<Component> {
  const nameMap: { [key: string]: number } = {};

  return components.map(component => {
    let name = component.name;

    if (nameMap[name]) {
      name += nameMap[name];
      nameMap[name]++;
    } else {
      nameMap[name] = 1;
    }

    return {
      ...component,
      name
    };
  });
}

function collectImports(
  components: ReadonlyArray<Component>
): ReadonlyArray<Import> {
  const importMap: { [key: string]: Set<string> } = {};

  for (const component of components) {
    for (const { path, names } of component.imports) {
      if (!importMap[path]) importMap[path] = new Set();

      for (const name of names) {
        importMap[path].add(name);
      }
    }
  }

  return Object.keys(importMap).map(path => ({
    path,
    names: [...importMap[path].values()]
  }));
}

export function migrate(data: ReadonlyArray<Manifest>): string {
  const components = uniqComponentName(generateForList(data));
  let output = `"use strict";\n\n`;

  for (const { path, names } of collectImports(components)) {
    output += `const { ${names.join(", ")} } = require("${path}");\n`;
  }

  for (const { name, text } of components) {
    output += `\nconst ${name} = ${text};\n`;
  }

  output += `\nmodule.exports = [${components.map(c => c.name).join(", ")}];\n`;

  return output;
}

export function migrateString(input: string) {
  return migrate(yaml.safeLoadAll(input));
}
