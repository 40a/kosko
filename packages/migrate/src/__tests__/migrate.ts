import { migrate, migrateString, Manifest } from "../migrate";
import { YAMLException } from "js-yaml";
import { runInThisContext } from "vm";
import Module from "module";

describe("migrate", () => {
  let data: ReadonlyArray<Manifest>;
  let code: string;
  let exported: any;

  function execute() {
    code = migrate(data);

    const mod = new Module("");
    const fn = runInThisContext(Module.wrap(code));
    fn(mod.exports, require, mod, "", "");
    exported = mod.exports;
  }

  describe("given no manifests", () => {
    beforeAll(() => {
      data = [];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given single manifest", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "v1",
          kind: "Pod",
          metadata: { name: "test-pod" },
          spec: {
            containers: []
          }
        }
      ];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given multiple manifests of different kinds", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "apps/v1",
          kind: "Deployment",
          metadata: { name: "test-deployment" },
          spec: {
            replicas: 2
          }
        },
        {
          apiVersion: "v1",
          kind: "Service",
          metadata: { name: "test-service" },
          spec: {
            selector: {}
          }
        }
      ];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given multiple manifests of same kind", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "v1",
          kind: "ConfigMap",
          metadata: {
            name: "config-foo"
          },
          data: {
            foo: "bar"
          }
        },
        {
          apiVersion: "v1",
          kind: "ConfigMap",
          metadata: {
            name: "config-bar"
          },
          data: {
            bar: "baz"
          }
        }
      ];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given a List", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "v1",
          kind: "List",
          items: [
            {
              apiVersion: "apps/v1",
              kind: "Deployment",
              metadata: { name: "test-deployment" },
              spec: {
                replicas: 2
              }
            },
            {
              apiVersion: "v1",
              kind: "Service",
              metadata: { name: "test-service" },
              spec: {
                selector: {}
              }
            }
          ]
        }
      ];
    });

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given a RBAC ClusterRoleBinding", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "rbac.authorization.k8s.io/v1",
          kind: "ClusterRoleBinding",
          metadata: {
            name: "tiller"
          }
        }
      ];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("given a CRD", () => {
    beforeAll(() => {
      data = [
        {
          apiVersion: "networking.istio.io/v1alpha3",
          kind: "VirtualService",
          metadata: { name: "details" }
        }
      ];
    });

    beforeEach(execute);

    test("should generate code", () => {
      expect(code).toMatchSnapshot();
    });

    test("should export value", () => {
      expect(exported).toMatchSnapshot();
    });
  });

  describe("when apiVersion does not exist", () => {
    test("should throw an error", () => {
      expect(() => migrate([{} as any])).toThrowError(
        "apiVersion and kind is required"
      );
    });
  });

  describe("when kind does not exist", () => {
    test("should throw an error", () => {
      expect(() => migrate([{} as any])).toThrowError(
        "apiVersion and kind is required"
      );
    });
  });
});

describe("migrateString", () => {
  describe("given valid YAML", () => {
    test("should generate code", () => {
      expect(
        migrateString(`---
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  containers: []
`)
      ).toMatchSnapshot();
    });
  });

  describe("given valid JSON", () => {
    test("should generate code", () => {
      expect(
        migrateString(`{
"apiVersion": "v1",
"kind": "Pod",
"metadata": {
  "name": "test-pod"
},
"spec": {
  "containers": []
}
      }`)
      ).toMatchSnapshot();
    });
  });

  describe("given invalid YAML", () => {
    test("should throw an error", () => {
      expect(() => migrateString("{")).toThrow(YAMLException);
    });
  });
});
