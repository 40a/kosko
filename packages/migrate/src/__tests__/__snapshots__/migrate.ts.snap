// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrate given a CRD should export an array containing a CRD 1`] = `
Array [
  Object {
    "apiVersion": "networking.istio.io/v1alpha3",
    "kind": "VirtualService",
    "metadata": Object {
      "name": "details",
    },
  },
]
`;

exports[`migrate given a CRD should generate code 1`] = `
"\\"use strict\\";


const virtualService = {
  \\"apiVersion\\": \\"networking.istio.io/v1alpha3\\",
  \\"kind\\": \\"VirtualService\\",
  \\"metadata\\": {
    \\"name\\": \\"details\\"
  }
};

module.exports = [virtualService];
"
`;

exports[`migrate given a List should export an array containing items in the list 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment",
    },
    "spec": Object {
      "replicas": 2,
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-service",
    },
    "spec": Object {
      "selector": Object {},
    },
  },
]
`;

exports[`migrate given a List should generate code 1`] = `
"\\"use strict\\";

const { Deployment } = require(\\"kubernetes-models/api/apps/v1/Deployment\\");
const { Service } = require(\\"kubernetes-models/api/core/v1/Service\\");

const deployment = new Deployment({
  \\"metadata\\": {
    \\"name\\": \\"test-deployment\\"
  },
  \\"spec\\": {
    \\"replicas\\": 2
  }
});

const service = new Service({
  \\"metadata\\": {
    \\"name\\": \\"test-service\\"
  },
  \\"spec\\": {
    \\"selector\\": {}
  }
});

module.exports = [deployment, service];
"
`;

exports[`migrate given a RBAC ClusterRoleBinding should export an array containing a ClusterRoleBinding 1`] = `
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "ClusterRoleBinding",
    "metadata": Object {
      "name": "tiller",
    },
  },
]
`;

exports[`migrate given a RBAC ClusterRoleBinding should generate code 1`] = `
"\\"use strict\\";

const { ClusterRoleBinding } = require(\\"kubernetes-models/api/rbac/v1/ClusterRoleBinding\\");

const clusterRoleBinding = new ClusterRoleBinding({
  \\"metadata\\": {
    \\"name\\": \\"tiller\\"
  }
});

module.exports = [clusterRoleBinding];
"
`;

exports[`migrate given multiple manifests of different kinds should export an array containing a deployment and a service 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment",
    },
    "spec": Object {
      "replicas": 2,
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-service",
    },
    "spec": Object {
      "selector": Object {},
    },
  },
]
`;

exports[`migrate given multiple manifests of different kinds should generate code 1`] = `
"\\"use strict\\";

const { Deployment } = require(\\"kubernetes-models/api/apps/v1/Deployment\\");
const { Service } = require(\\"kubernetes-models/api/core/v1/Service\\");

const deployment = new Deployment({
  \\"metadata\\": {
    \\"name\\": \\"test-deployment\\"
  },
  \\"spec\\": {
    \\"replicas\\": 2
  }
});

const service = new Service({
  \\"metadata\\": {
    \\"name\\": \\"test-service\\"
  },
  \\"spec\\": {
    \\"selector\\": {}
  }
});

module.exports = [deployment, service];
"
`;

exports[`migrate given multiple manifests of same kind should export an array containing three config maps 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "foo": "bar",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "config-foo",
    },
  },
  Object {
    "apiVersion": "v1",
    "data": Object {
      "bar": "baz",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "config-bar",
    },
  },
  Object {
    "apiVersion": "v1",
    "data": Object {
      "baz": "boo",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "config-baz",
    },
  },
]
`;

exports[`migrate given multiple manifests of same kind should generate code 1`] = `
"\\"use strict\\";

const { ConfigMap } = require(\\"kubernetes-models/api/core/v1/ConfigMap\\");

const configMap = new ConfigMap({
  \\"metadata\\": {
    \\"name\\": \\"config-foo\\"
  },
  \\"data\\": {
    \\"foo\\": \\"bar\\"
  }
});

const configMap1 = new ConfigMap({
  \\"metadata\\": {
    \\"name\\": \\"config-bar\\"
  },
  \\"data\\": {
    \\"bar\\": \\"baz\\"
  }
});

const configMap2 = new ConfigMap({
  \\"metadata\\": {
    \\"name\\": \\"config-baz\\"
  },
  \\"data\\": {
    \\"baz\\": \\"boo\\"
  }
});

module.exports = [configMap, configMap1, configMap2];
"
`;

exports[`migrate given no manifests should generate code 1`] = `
"\\"use strict\\";


module.exports = [];
"
`;

exports[`migrate given single manifest should export an array containing a pod 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "test-pod",
    },
    "spec": Object {
      "containers": Array [],
    },
  },
]
`;

exports[`migrate given single manifest should generate code 1`] = `
"\\"use strict\\";

const { Pod } = require(\\"kubernetes-models/api/core/v1/Pod\\");

const pod = new Pod({
  \\"metadata\\": {
    \\"name\\": \\"test-pod\\"
  },
  \\"spec\\": {
    \\"containers\\": []
  }
});

module.exports = [pod];
"
`;

exports[`migrateString given valid JSON should generate code 1`] = `
"\\"use strict\\";

const { Pod } = require(\\"kubernetes-models/api/core/v1/Pod\\");

const pod = new Pod({
  \\"metadata\\": {
    \\"name\\": \\"test-pod\\"
  },
  \\"spec\\": {
    \\"containers\\": []
  }
});

module.exports = [pod];
"
`;

exports[`migrateString given valid YAML should generate code 1`] = `
"\\"use strict\\";

const { Pod } = require(\\"kubernetes-models/api/core/v1/Pod\\");

const pod = new Pod({
  \\"metadata\\": {
    \\"name\\": \\"test-pod\\"
  },
  \\"spec\\": {
    \\"containers\\": []
  }
});

module.exports = [pod];
"
`;
